modules/EvManager/main/car_simulatorImpl.cpp

```diff
@@ -0,0 +1,556 @@
+
+#include "car_simulatorImpl.hpp"
+#include "main/SimCommand.hpp"
+#include <everest/logging.hpp>
+#include <queue>
+
+namespace module::main {
+
+void car_simulatorImpl::init() {
+    // TODO: simulation loop time from manifest.yaml?
+    simData = std::make_unique<main::SimData>();
+    subscribeToExternalMQTT();
+    subscribeToVariablesOnInit();
+}
+
+void car_simulatorImpl::ready() {
+    registerAllCommands();
+    enabled = false;
+
+    setupEVParameters();
+
+    if (mod->config.auto_enable) {
+        auto enableCopy = mod->config.auto_enable;
+        handle_enable(enableCopy);
+    }
+    if (mod->config.auto_exec) {
+        auto valueCopy = mod->config.auto_exec_commands;
+        handle_executeChargingSession(valueCopy);
+    }
+}
+
+void car_simulatorImpl::handle_enable(bool& value) {
+    if (!enabled.has_value()) {
+        EVLOG_warning << "Already received data, but framework is not ready yet.";
+        return;
+    }
+
+    if (enabled == value) {
+        // ignore if value is the same
+        return;
+    }
+
+    simData = std::make_unique<main::SimData>();
+
+    callEVBoardSupportFunctions();
+
+    // set loop interval
+    if (value) {
+        enabled = true;
+        loopIntervalMs = defaultLoopIntervalMs;
+        simulationThread = std::thread{[this] { handleSimulationLoop(); }};
+    } else {
+        enabled = false;
+        loopIntervalMs.reset();
+        simulationThread.join();
+    }
+
+    mod->r_ev_board_support->call_enable(value);
+    publish_enabled(value);
+}
+
+void car_simulatorImpl::handle_executeChargingSession(std::string& value) {
+    // Check enabled
+    if (!checkCanExecute()) {
+        return;
+    }
+
+    simData = std::make_unique<main::SimData>();
+
+    auto valueCopy = value;
+    auto& commandQueue = simData->commandQueue;
+    commandQueue = parseSimCommands(valueCopy);
+
+    // Start execution
+    if (!commandQueue.empty()) {
+        simData->executionActive = true;
+    }
+}
+
+void car_simulatorImpl::handle_modifyChargingSession(std::string& value) {
+    if (!enabled) {
+        EVLOG_warning << "Simulation disabled, cannot execute charging simulation.";
+        return;
+    }
+
+    auto valueCopy = value;
+    simData->commandQueue = parseSimCommands(valueCopy);
+
+    if (!simData->commandQueue.empty()) {
+        simData->executionActive = true;
+    }
+}
+
+void car_simulatorImpl::handleSimulationLoop() {
+    while (enabled) {
+        if (simData && simData->executionActive) {
+            if (loopIntervalMs.has_value()) {
+                runSimulationLoop();
+                std::this_thread::sleep_for(std::chrono::milliseconds(loopIntervalMs.value()));
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+void car_simulatorImpl::registerAllCommands() {
+    const auto sleepFuntion = [this](const std::vector<std::string>& arguments) {
+        static std::optional<size_t> sleepTimeLeftMs;
```

Comment:
I would consider using this static variable here a bit hacky

Reply:
You are right. At the time I was thinking there will only be one instance of the `car_simulatorImpl` running so this would make sense but I guess there could be multiple. I will change that